Admin/Test-AMSI.ps1

```diff
@@ -1,144 +1,813 @@
+
+    Write-Host ("Test-AMSI.ps1 script version $($BuildVersion)") -ForegroundColor Green
+
+    if ($ScriptUpdateOnly) {
+        switch (Test-ScriptVersion -AutoUpdate -VersionsUrl "https://aka.ms/Test-AMSI-VersionsURL" -Confirm:$false) {
+        ($true) { Write-Host ("Script was successfully updated") -ForegroundColor Green }
+        ($false) { Write-Host ("No update of the script performed") -ForegroundColor Yellow }
+            default { Write-Host ("Unable to perform ScriptUpdateOnly operation") -ForegroundColor Red }
+        }
         return
     }
-    if ($DisableAMSI) {
-        New-SettingOverride -Name DisablingAMSIScan -Component Cafe -Section HttpRequestFiltering -Parameters ("Enabled=False") -Reason "Disabled via CSS-Exchange Script"
-        Get-ExchangeDiagnosticInfo -Process Microsoft.Exchange.Directory.TopologyService -Component VariantConfiguration -Argument Refresh
-        Write-Warning "Remember to restart IIS for this to take affect. You can accomplish this by running .\Test-AMSI.ps1 -RestartIIS"
+
+    if ((-not($SkipVersionCheck)) -and
+    (Test-ScriptVersion -AutoUpdate -VersionsUrl "https://aka.ms/Test-AMSI-VersionsURL" -Confirm:$false)) {
+        Write-Host ("Script was updated. Please re-run the command") -ForegroundColor Yellow
         return
     }
-    if ($RestartIIS) {
-        Restart-Service -Name W3SVC, WAS -Force
-        return
+
+    $msgNewLine = "`n"
+    if (-not (Confirm-Administrator)) {
+        Write-Host $msgNewLine
+        Write-Warning "This script needs to be executed in elevated mode. Start the Exchange Management Shell as an Administrator and try again."
+        exit
     }
-}
 
-Test-AMSI
+    $exchangeShell = Confirm-ExchangeShell
+    if (-not($exchangeShell.ShellLoaded)) {
+        Write-Host $msgNewLine
+        Write-Warning "Failed to load Exchange Shell Module..."
+        exit
+    }
+
+    if ((Get-ExchangeServer $env:COMPUTERNAME).IsEdgeServer) {
+        Write-Host $msgNewLine
+        Write-Warning "This script cannot be executed in an Edge Server."
+        exit
+    }
+
+    $bar = ""
+    1..($Host.UI.RawUI.WindowSize.Width) | ForEach-Object { $bar += "-" }
+    Write-Host ""
+    Write-Host $bar
+
+    $filterList = @()
+
+    if ($PSCmdlet.ParameterSetName -eq "TestAMSI" -or $PSCmdlet.ParameterSetName -eq "TestAMSIAll") {
+        $TestAMSI = $true
+    }
+
+    if ($null -eq $ServerList -and ($RestartIIS -or $CheckAMSIConfig -or $TestAMSI) -and (-not $AllServers)) {
+        $ServerList = $env:COMPUTERNAME
+    }
+
+    Write-Host ""
+    Write-Host "AMSI only support on Exchange 2016 CU21 or newer, Exchange 2019 CU10 or newer and running on Windows 2016 or newer" -ForegroundColor Green
+    Write-Host ""
+
+    $SupportedExchangeServers = Get-ExchangeServer | Where-Object {
+        ($_.IsClientAccessServer) -and
+        ((((Get-ExchangeBuildVersionInformation -AdminDisplayVersion ($_.adminDisplayVersion.ToString())).BuildVersion.Minor -eq 1) -and
```

Comment:
I think this could be simplified (maybe like this - have not tested)

```
$exchangeServer = New-Object 'System.Collections.Generic.List[object]'
Get-ExchangeServer | Where-Object { $_.IsClientAccessServer } | ForEach-Object {
    $server = $_
    $versionInformation = Get-ExchangeBuildVersionInformation -AdminDisplayVersion $_.AdminDisplayVersion
    switch ($versionInformation.MajorVersion) {
        "Exchange2016" { if ($versionInformation.BuildVersion -ge "15.1.2308.8") { $exchangeServer.Add($server); break } }
        "Exchange2019" { if ($versionInformation.BuildVersion -ge "15.2.922.7") { $exchangeServer.Add($server); break } }
    }
}
```